testing/soak/src/integTest/groovy/org/gradle/cache/LockCommunicationSoakTest.groovy

```diff
@@ -0,0 +1,296 @@
+        def oldHandle = oldExecuter.withTasks('lock').start()
+        aboutToRequest.waitForAllPendingCalls()
+
+        // new Gradle should be holding the lock
+
+        aboutToRequest.releaseAll()
+
+        // The old Gradle is now waiting to acquire the lock
+        // New Gradle can now release the lock
+        releaseLock.releaseAll()
+
+        // Old Gradle should acquire the lock
+        acquireLock.waitForAllPendingCalls()
+        acquireLock.releaseAll()
+
+        newHandle.waitForFinish()
+        oldHandle.waitForFinish()
+    }
+
+    def "lock contention - happy path old to new"() {
+        def aboutToLock = blockingServer.expectAndBlock("old-first")
+        def releaseLock = blockingServer.expectAndBlock("old-second")
+        def aboutToRequest = blockingServer.expectAndBlock("new-first")
+        def acquireLock = blockingServer.expectAndBlock("new-second")
+
+        expect:
+        def oldHandle = oldExecuter.withTasks('lock').start()
+
+        aboutToLock.waitForAllPendingCalls()
+        aboutToLock.releaseAll()
+        releaseLock.waitForAllPendingCalls()
+
+        // Start the new Gradle to request the same lock
+        def newHandle = executer.withTasks('lock').start()
+        aboutToRequest.waitForAllPendingCalls()
+
+        // old Gradle should be holding the lock
+
+        aboutToRequest.releaseAll()
+
+        // The new Gradle is now waiting to acquire the lock
+        // Old Gradle can now release the lock
+        releaseLock.releaseAll()
+
+        // New Gradle should acquire the lock
+        acquireLock.waitForAllPendingCalls()
+        acquireLock.releaseAll()
+
+        newHandle.waitForFinish()
+        oldHandle.waitForFinish()
+    }
+
+    def "lock contention - poison new"() {
+        def aboutToLock = blockingServer.expectAndBlock("new-first")
+        def releaseLock = blockingServer.expectAndBlock("new-second")
+        def aboutToRequest = blockingServer.expectAndBlock("old-first")
+        def acquireLock = blockingServer.expectAndBlock("old-second")
+
+        expect:
+        def newHandle = executer.requireIsolatedDaemons().withTasks('lock').start()
+
+        aboutToLock.waitForAllPendingCalls()
+        aboutToLock.releaseAll()
+        releaseLock.waitForAllPendingCalls()
+
+        // Start the old Gradle to request the same lock
+        def oldHandle = oldExecuter.withTasks('lock').start()
+        aboutToRequest.waitForAllPendingCalls()
+
+        // new Gradle should be holding the lock
+        poison(file("port").text.toInteger())
+
+        aboutToRequest.releaseAll()
+
+        // The old Gradle is now waiting to acquire the lock
+        // New Gradle can now release the lock
+        releaseLock.releaseAll()
+
+        // Old Gradle should acquire the lock
+        acquireLock.waitForAllPendingCalls()
+        acquireLock.releaseAll()
+
+        newHandle.waitForFinish()
+        oldHandle.waitForFinish()
+    }
+
+    def "lock contention - poison old"() {
```

Comment:
🤔 I don't understand this test tbh

The old Gradle should have its listener thread die when poisoned, but the tests finishes with a successful path. What am I missing?

Reply:
Blah, I was trying to make the tests look similar and fail consistently, but at some point, I made this one pass. 

What was happening is the old Gradle was exiting (and releasing locks), so the new Gradle could get the lock.

I've redone this to show what will happen when old Gradle gets into a bad state. New Gradle should fail.